ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/PullQueryExecutor.java

```diff
@@ -224,26 +230,48 @@ public PullQueryResult execute(
           .getMaterialization(queryId, contextStacker)
           .orElseThrow(() -> notMaterializedException(getSourceName(analysis)));
 
-      final Struct key = asKeyStruct(whereInfo.keyBound, query.getPhysicalSchema());
+      final List<Optional<KsqlNode>> sourceNodes = new ArrayList<>();
+      final List<List<?>> tableRows = new ArrayList<>();
+      final List<LogicalSchema> schemas = new ArrayList<>();
+      final List<Future<PullQueryResult>> futures = new ArrayList<>();
+      final List<List<Struct>> keysByLocation = mat.locator().groupByLocation(
+          whereInfo.keysBound.stream()
+              .map(keyBound -> asKeyStruct(keyBound, query.getPhysicalSchema()))
+              .collect(Collectors.toList()));
+
+      for (List<Struct> keys : keysByLocation) {
+        final PullQueryContext pullQueryContext = new PullQueryContext(
```

Comment:
The `pullQueryContext` is the same, the only thing that changes are the keys right?You could avoid creating a new object and just set the keys in each iteration.

Reply:
`PullQueryContext` is threadsafe because it's immutable.  Dealing with small immutable objects is fairly efficient in the JVM and is considered worth doing to make things threadsafe and easy to understand.